feat: add blacklisted block check

matthewkeil opened the pull request:
**Motivation**

Try to save Holesky.  Blacklist ancestor of invalid chain and add blacklist BlockError that can be expanded to a feature flag in the future (or API call to add to the errors array)

matthewkeil committed:
feat: add blacklisted block check

matthewkeil committed:
feat: add cli flag for blacklistedBlocks

matthewkeil committed:
fix: remove hard-coded bad block

matthewkeil committed:
fix: lint and bug in cli coerce

matthewkeil committed:
fix: dry ensure0xPrefix

wemeetagain commented on packages/cli/src/options/beaconNodeOptions/chain.ts:
```diff
@@ -106,6 +109,19 @@ export const options: CliCommandOptions<ChainArgs> = {
+    hidden: true,
+    type: "array",
+    description:
+      "Common separated list of 0x prefixed root hex's for blocks that should not be allowed through processing",
```
```suggestion
      "Comma-separated list of 0x-prefixed root hex's for blocks that should not be allowed through processing",
```

twoeths committed:
feat: black list the whole chain leading to black listed blocks (#7500)

**Motivation**

- Black list the whole chain leading to black listed blocks

**Description**

- Init chain's black listed blocks from chain options
- Populate it if a block's parent root is black listed. Gradually the
whole chain will be black listed

---------

Co-authored-by: Tuyen Nguyen <twoeths@users.noreply.github.com>
Co-authored-by: matthewkeil <me@matthewkeil.com>

ensi321 reviewed (commented):
Is this intended to be merged into unstable or is it only to showcase this feature in holesky-rescue? 

If former, I don't really think we should pass the blacklisted blocks from cli arguments but probably read it from a file since it can get lengthy.

There is also an argument brought up by lighthouse guys [here](https://github.com/sigp/lighthouse/pull/7042/files/05e0871dd47271fd4aa434bcf1d12eb89ef982fd#r1971614634) about this flag can potentially be abused to censor blocks, so we need to take this into account when making design decisions eg. make it harder to use, limit number of blacklisted blocks, restrict the this feature on mainnet etc.

nflaig commented:
> flag can potentially be abused to censor blocks

I don't see how that is possible, the node needs to be restarted to black list a block but at this point the block should already have enough votes to be considered part of the canonical chain and the node that blacklists it would just follow a separate (minority) chain. I can't really think of a scenario here, but maybe I am missing something obvious

philknows commented:
As discussed on standup #7524 , if this feature is going ahead, it should be merged to unstable before Friday deadline for v1.28-rc.0

nflaig committed:
Merge branch 'unstable' into mkeil/holesky-rescue

nflaig reviewed (dismissed):
LGTM

nflaig committed:
Lint

nflaig commented:
> we can also add an api returning slot/root hexes of black listed blocks

@twoeths this is a comment from you on discord, do you feel like we still want this? seems simple to implement but I don't feel strongly about adding this

although on second thought it might be helpful to see all black listed parent blocks

twoeths commented:
> > we can also add an api returning slot/root hexes of black listed blocks
> 
> @twoeths this is a comment from you on discord, do you feel like we still want this? seems simple to implement but I don't feel strongly about adding this
> 
> although on second thought it might be helpful to see all black listed parent blocks

I think we should add it considering how easy it is to implement
during devnets (hopefully not mainnet), if issues happen we can easily debug, that would be a critical input for the devnet to investigate. For example from black listed blocks we can trace which clients sent us that...

wemeetagain commented:
:tada: This PR is included in [v1.28.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.28.0) :tada:
